kong/utils.go

```diff
@@ -208,9 +214,113 @@ func fillConfigRecord(schema gjson.Result, config Configuration) Configuration {
 	return res
 }
 
+// flattenDefaultsSchema gets an arbitrarily nested and structured entity schema
+// and flattens it, turning it into a map that can be more easily unmarshalled
+// into proper entity objects.
+//
+// Sample input:
+// {
+// 	"fields": [
+//         {
+//             "algorithm": {
+//                 "default": "round-robin",
+//                 "one_of": ["consistent-hashing", "least-connections", "round-robin"],
+//                 "type": "string"
+//             }
+//         }, {
+//             "hash_on": {
+//                 "default": "none",
+//                 "one_of": ["none", "consumer", "ip", "header", "cookie"],
+//                 "type": "string"
+//             }
+//         }, {
+//             "hash_fallback": {
+//                 "default": "none",
+//                 "one_of": ["none", "consumer", "ip", "header", "cookie"],
+//                 "type": "string"
+//             }
+//         },
+//   ...
+// }
+//
+// Sample output:
+// {
+// 	"algorithm": "round-robin",
+// 	"hash_on": "none",
+// 	"hash_fallback": "none",
+//  ...
+// }
+func flattenDefaultsSchema(schema gjson.Result) Schema {
+	value := schema.Get("fields")
+	results := Schema{}
+
+	value.ForEach(func(key, value gjson.Result) bool {
+		// get the key name
+		ms := value.Map()
+		fname := ""
+		for k := range ms {
+			fname = k
+			break
+		}
+
+		ftype := value.Get(fname + ".type")
+		if ftype.String() == "record" {
+			newSubConfig := flattenDefaultsSchema(value.Get(fname))
+			results[fname] = newSubConfig
+			return true
+		}
+		value = value.Get(fname + ".default")
+		if value.Exists() {
+			results[fname] = value.Value()
+		} else {
+			results[fname] = nil
+		}
+		return true
+	})
+
+	return results
+}
+
+func getDefaultsObj(schema Schema) ([]byte, error) {
+	jsonSchema, err := json.Marshal(&schema)
+	if err != nil {
+		return nil, err
+	}
+	gjsonSchema := gjson.ParseBytes((jsonSchema))
+	defaults := flattenDefaultsSchema(gjsonSchema)
+	jsonSchemaWithDefaults, err := json.Marshal(&defaults)
+	if err != nil {
+		return nil, err
+	}
+	return jsonSchemaWithDefaults, nil
+}
+
+// FillEntityDefaults ingests entities' defaults from their schema.
+func FillEntityDefaults(entity interface{}, schema Schema) error {
+	var tmpEntity interface{}
+	switch entity.(type) {
+	case *Target:
+		tmpEntity = &Target{}
+	case *Service:
+		tmpEntity = &Service{}
+	case *Route:
+		tmpEntity = &Route{}
+	case *Upstream:
+		tmpEntity = &Upstream{}
+	}
+	defaults, err := getDefaultsObj(schema)
```

Comment:
check for schema == nil before using

Reply:
Mmmh what do you think it's best here if `schema == nil`, returning `nil` or erroring out complaining about it being `nil`? Maybe the former?

Reply:
Return an error. When in doubt, return an error to the caller. That leads to better UX because the caller can then appropriate action. If the code returns without an error, the caller may assume that the function call worked only to be surprised later that it didn't.

Reply:
roger that